src/app/AttributeAccessInterface.h

```diff
@@ -41,23 +42,116 @@
+        {
+            return mAttributeValueEncoder->EncodeListItem(std::forward<Ts>(aArgs)...);
+        }
+
+    private:
+        AttributeValueEncoder * mAttributeValueEncoder;
+    };
+
+    class AttributeEncodeState
+    {
+    public:
+        AttributeEncodeState() = default;
+
+        bool mAllowPartialData            = false;
+        ListIndex mCurrentEncodeListIndex = kInvalidListIndex;
+    };
+
+    AttributeValueEncoder(AttributeReportIB::Builder & aAttributeReportIBBuilder, FabricIndex aAccessingFabricIndex,
+                          const ConcreteAttributePath & aPath, const AttributeEncodeState & aState = { false, kInvalidListIndex }) :
+        mAttributeReportIBBuilder(aAttributeReportIBBuilder),
+        mAccessingFabricIndex(aAccessingFabricIndex), mPath(aPath), mEncodeState(aState)
     {}
 
     template <typename... Ts>
     CHIP_ERROR Encode(Ts... aArgs)
     {
         mTriedEncode = true;
-        if (mWriter == nullptr)
+
+        AttributeDataIB::Builder attributeDataIBBuilder = mAttributeReportIBBuilder.CreateAttributeData();
+        AttributePathIB::Builder attributePathIBBuilder = attributeDataIBBuilder.CreatePath();
+        attributePathIBBuilder.Endpoint(mPath.mEndpointId).Cluster(mPath.mClusterId).Attribute(mPath.mAttributeId);
+
+        if (mCurrentEncodingListIndex != kInvalidListIndex)
+        {
+            attributePathIBBuilder.ListIndex(mCurrentEncodingListIndex);
+        }
+
+        attributePathIBBuilder.EndOfAttributePathIB();
+
+        ReturnErrorOnFailure(
+            TagBoundEncoder(attributeDataIBBuilder.GetWriter(), TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData)))
+                .Encode(std::forward<Ts>(aArgs)...));
+
+        attributeDataIBBuilder.DataVersion(0).EndOfAttributeDataIB();
+        return attributeDataIBBuilder.GetError();
+    }
+
+    template <typename... Ts>
+    CHIP_ERROR EncodeListItem(Ts... aArgs)
+    {
+        if (mEncodeState.mCurrentEncodeListIndex == kInvalidListIndex)
         {
+            AttributeDataIB::Builder attributeDataIBBuilder = mAttributeReportIBBuilder.CreateAttributeData();
+            attributeDataIBBuilder.CreatePath()
+                .Endpoint(mPath.mEndpointId)
+                .Cluster(mPath.mClusterId)
+                .Attribute(mPath.mAttributeId)
+                .EndOfAttributePathIB();
+
+            ReturnErrorOnFailure(
+                TagBoundEncoder(attributeDataIBBuilder.GetWriter(), TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData)))
+                    .EncodeList([](const TagBoundEncoder &) -> CHIP_ERROR { return CHIP_NO_ERROR; }));
+
+            attributeDataIBBuilder.DataVersion(0).EndOfAttributeDataIB();
+            ReturnErrorOnFailure(attributeDataIBBuilder.GetError());
+
+            mCurrentEncodingListIndex            = 0;
+            mEncodeState.mCurrentEncodeListIndex = 0;
+        }
+
+        if (mCurrentEncodingListIndex < mEncodeState.mCurrentEncodeListIndex)
```

Comment:
I'd really re-name `mCurrentEncodeListIndex` in the mEncodeState to `mTargetEncodeListIndex` to at least provide some distinction in the names.